### ISSUE TITLE
Invalid IDE0059 Suggestion

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#44010]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/1015839/invalid-ide0059-suggestion.html)._

---
I have the following C# method.

The second to last statement (`parser++;`) gives me an invalid *IDE0059 Unecessary assignment of a value to 'parser'* suggestion.

```
/// <summary>
/// Evaluates each parameter of a function's parameter list and returns
/// a list of those values. An empty list is returned if no parameters
/// were found. It is assumed the current position is at the opening
/// parenthesis of the argument list.
/// </summary>
/// <param name="parser">ParsingHelper object.</param>
/// <returns>A list of parameter values.</returns>
private Variable[] ParseParameters(ParsingHelper parser)
{
    List<Variable> parameters = new List<Variable>();

// Move past open parenthesis
    parser++;
    parser. SkipWhiteSpace();
    // If function has any parameters
    if (parser. Peek() != ')')
    {
        // Parse function parameter list
        int start = parser. Index;
        int parenCount = 1;

while (!parser. EndOfText)
        {
            if (parser. Peek() == ',')
            {
                // Note: Ignore commas inside parentheses. They could be
                // for a parameter list in a function inside the parameters
                if (parenCount == 1)
                {
                    parameters. Add(ParseParameter(parser, start));
                    start = parser. Index + 1;
                }
            }
            if (parser. Peek() == ')')
            {
                parenCount--;
                if (parenCount == 0)
                {
                    parameters. Add(ParseParameter(parser, start));
                    break;
                }
            }
            else if (parser. Peek() == '(')
            {
                parenCount++;
            }
            parser++;
        }
    }
    // Make sure we found a closing parenthesis
    if (parser. Peek() != ')')
        throw new ExpressionException(ErrClosingParenExpected, parser. Index);
    // Move past closing parenthesis
    parser++;
    // Return parameter list
    return parameters. ToArray();
}
```

At first glance, this seems correct because I'm apparently incrementing the value just before the method returns.

However, this class overloads the `++` operator to increment an internal counter. Therefore removing the statement would change the functionality.

```
public static ParsingHelper operator ++(ParsingHelper helper);
```

---
#### Visual Studio Feedback System on 5/4/2020, 02:05 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by MisinformedDNA** [comment]
This is because you aren't using `parser` after `parser++`, so the assignment is indeed unneeded.

@sharwell Propose close.

**Comment by SoftCircuits** [comment]
@MisinformedDNA Looks like you didn't read to the end. It is used again by the calling method. Please withdraw your proposal to close unless you read the entire description.

**Comment by MisinformedDNA** [comment]
I don't have time to read. Lol.

"Proposed close" removed. Can you share a complete, minimal repro with all the essential code?

**Comment by SoftCircuits** [comment]
@MisinformedDNA Well, not sure why it's needed. It's really straight forward. But here's a complete program that demonstrates the suggestion.

``` cs
Parser parser = new();
DoParsing(parser);
Console.WriteLine(parser.Index);

static void DoParsing(Parser parser)
{
    parser++;    // IDE0059 Unnecessary assignment of a value to 'parser'
}

class Parser
{
    public int Index;

    public Parser()
    {
        Index = 0;
    }

    public static Parser operator ++(Parser parser)
    {
        parser.Index++;
        return parser;
    }
}
```

**Comment by MisinformedDNA** [comment]
Explanation:

There are two references pointing to the same object.

```cs
Parser parser = new();  // `parser` is a reference to the created object

DoParsing(Parser parser) // This `parser` is a second reference to the created object
```

`parser++` assigns itself to the second reference. Since the second reference is not used, the assignment is "unnecessary".

However, in the overload, you are "deferencing the object" and changing the object itself, so it works as you expect.

To remove the second reference, you need the `ref` keyword

```cs
static void DoParsing(ref Parser parser)
```

I think the analyzer is working as intended.

**Comment by CyrusNajmabadi** [comment]
@MisinformedDNA Parser is a class.  It doesn't need to be passed by ref.  `++` is mutating the internal state of the single class object.

**Comment by MisinformedDNA** [comment]
Right, but `parser++` is essentially `parser = ++(parser)`, right? Then the analyzer will see that as getting a new reference. To cover this case, the analyzer would need to analyze the overload to determine whether it was returning a new reference or the same reference. It's up to you guys how you want to handle it but using `ref` does get rid of the warning.

**Comment by CyrusNajmabadi** [comment]
> Then the analyzer will see that as getting a new reference.

The analyzer can also be conservative here, and not make assumptions on what an overloaded `++` operator does.

**Comment by SoftCircuits** [comment]
@MisinformedDNA I have no idea why you think the number of references to the class is important here.

The bottom line is that the suggestion is telling me a line has no effect when in fact the line does have an effect. And the other issues you're raising doesn't address that.

**Comment by MisinformedDNA** [comment]
There is definitely an effect, it's the implicit assignment that is triggering the warning. 

@CyrusNajmabadi Can I ask if my explanation of the references is correct? And if not, what is wrong? You mentioned not needing `ref` with classes, but that is a C# feature. :)

**Comment by SoftCircuits** [comment]
> @CyrusNajmabadi Can I ask if my explanation of the references is correct? And if not, what is wrong? You mentioned not needing `ref` with classes, but that is a C# feature. :)

The ref keyword is only needed if you want the reference to be updateable. The keyword is not needed if you only want to be able to update the referenced object. This is the same in VB.NET.

**Comment by CyrusNajmabadi** [comment]
> Can I ask if my explanation of the references is correct? And if not, what is wrong? 

Your explanation is not correct.  Use of `ref` is only needed if you want to update the LValue (the location) passed into the method.  But that's not necessary here.  Here, it's not the location that is being updated, but the object itself.  

@SoftCircuits' code is not common, but is legal, and the analyzer is makign an errant assumption that is causing a suggestion that breaks it.
